### PR TITLE
bib: use container based buildroot when building bootc disk images (HMS-3318)

### PR DESCRIPTION
**Pull request by mvo5**
Instead of using the trandition rpm based buildroot this commit
switches the buildroot to the target bootc container. The rational
is that the low-level tooling like mkfs etc should match the target
system.

Once that has landed and we can use container based buildroots
I will start working on `bootc install-filesystem` instead of the ostree
pipeline.

### CODE REVIEW

File: Containerfile
Comment (cgwalters) [inline Containerfile:2]: Looks like you ran into https://github.com/osbuild/bootc-image-builder/pull/126

**Comment by mvo5** [comment]
Fwiw, I tested that centos-9 can be build/booted with this branch:
```
$ git diff
diff --git a/test/containerbuild.py b/test/containerbuild.py
index 96b24fc..2b289ba 100644
--- a/test/containerbuild.py
+++ b/test/containerbuild.py
@@ -24,5 +24,5 @@ def container_to_build_ref():
     # making "image_type" an "image" tuple (type, container_ref_to_test)
     return os.getenv(
         "BIB_TEST_BOOTC_CONTAINER_TAG",
-        "quay.io/centos-bootc/fedora-bootc:eln",
+        "quay.io/centos-bootc/centos-bootc-dev:stream9",
     )
$ sudo pytest-3 -s -rs -vv './test/test_build.py::test_image_boots[qcow2]'
[    0.000000] Linux version 5.14.0-407.el9.x86_64 (mockbuild@x86-05.stream.rdu4
...
[  OK  ] Finished Record Runlevel Change in UTMP.
vm ready at port 36639
hello
PASSED
```
the "regular" version seems to be working now as well, it seems we got some updates here (where can I see those? is there some RSS, mailing list or simialr?), I will prepare updated tests in a followup (or here if that is prefered).

**Comment by achilleas-k** [comment]
Tested locally too and on macOS and it's all looking good.  Nice work!

Some things I see in the build log that we might want to look into (perhaps in a follow-up).

1. The selinux stage is complaining about conflicting specifications:
```
org.osbuild.selinux: 44b959365d4586c0fc428fb0701b0b3077b9824cfcf6c9867298d33cb2b953de {
  "file_contexts": "etc/selinux/targeted/contexts/files/file_contexts",
  "labels": {
    "/usr/bin/ostree": "system_u:object_r:install_exec_t:s0"
  }
}
setfiles: Regex version mismatch, expected: 10.42 2022-12-11 actual: 10.40 2022-04-14
setfiles: Regex version mismatch, expected: 10.42 2022-12-11 actual: 10.40 2022-04-14
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/0ef9ada2ee87792e8ba21afd65aa00d79a1253018832652b8694862fb80e84.file and /run/osbuild/tree/usr/lib/firmware/cirrus/cs35l41-dsp1-spk-prot-103c8b8f-r1.bin.xz, using system_u:object_r:lib_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/13aa01655b8dec2a33725304d36c4f71c3f27b7c3218c80201488a57c956a9.file and /run/osbuild/tree/usr/lib/firmware/kaweth/trigger_code.bin.xz, using system_u:object_r:lib_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/2f94bd8f4dc18756df37ec6311ff462d1a2a38a6a2b11827884bf39750ddf3.file and /run/osbuild/tree/usr/lib/firmware/amdgpu/navi14_pfp.bin.xz, using system_u:object_r:lib_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/478c4c5ba6939502c3cab7efe21bad6eeff329a40341e4bcc1e6a3ca36405a.file and /run/osbuild/tree/usr/lib/firmware/intel/catpt/bdw/dsp_basefw.bin.xz, using system_u:object_r:lib_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/4a1cc69efc6607d55810d31a95c69c21169f73100b3b90adfc132b67c7c422.file and /run/osbuild/tree/usr/lib/firmware/nvidia/tu104/gr/gpccs_sig.bin.xz, using system_u:object_r:lib_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/4bc17d1d8a724d1766ce058897cd77db5581e64c81dcd1bd650f68f6211ed6.file and /run/osbuild/tree/usr/lib/firmware/qed/qed_init_values_zipped-8.37.2.0.bin.xz, using system_u:object_r:lib_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/a1fc086be45a8afb18020a84b336808724459354f94ec30df381855baa9b74.file and /run/osbuild/tree/usr/lib/firmware/mwl8k/fmimage_8687.fw.xz, using system_u:object_r:lib_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/e6263242972474d12095c5c2db24bc91d6a451d505aadbab72b877aa39bb33.file and /run/osbuild/tree/usr/lib/firmware/go7007/s2250-1.fw.xz, using system_u:object_r:lib_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/01c633b255fbba0ccf606435a1d28ad4e77232f78deb9b3489b5312dea182e.file and /run/osbuild/tree/usr/share/man/man1/podman-container-export.1.gz, using system_u:object_r:man_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/9225596df2f98a2e465cd98642fa1bf4f321f54db114384416320047d91872.file and /run/osbuild/tree/usr/lib/modules/5.14.0-407.el9.x86_64/kernel/sound/soc/codecs/snd-soc-max98357a.ko.xz, using system_u:object_r:modules_object_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/bd7d6973a5360a07cd8e463e1cc6a8e61ee88071f55244b8a303bef6ce0757.file and /run/osbuild/tree/usr/lib/modules/5.14.0-407.el9.x86_64/kernel/drivers/usb/misc/iowarrior.ko.xz, using system_u:object_r:modules_object_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/111c9d3a033202ab976953d0bba275528b4632c28eec2830b9720545d20155.file and /run/osbuild/tree/usr/lib64/python3.9/xml/dom/__pycache__/domreg.cpython-39.opt-2.pyc, using system_u:object_r:lib_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/1941ddbbc2d92241aec60a3898f0128e3c4b741b4bab7e97d0af1aceffe79a.file and /run/osbuild/tree/usr/lib64/python3.9/multiprocessing/popen_forkserver.py, using system_u:object_r:lib_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/1974201cae276307d8adc3d0e2e3e443396f6ddb1a2dbbd61c8184f1149024.file and /run/osbuild/tree/usr/lib64/python3.9/__pycache__/stat.cpython-39.opt-2.pyc, using system_u:object_r:lib_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/479986eb4368929493325c669914cc40798d9bf2d8ab5e110d7295795ef17a.file and /run/osbuild/tree/usr/lib64/python3.9/xml/dom/__pycache__/xmlbuilder.cpython-39.pyc, using system_u:object_r:lib_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/559f7672284d46b02fc685ae6750895789f455f71bb7f2c06de5f4e83c9375.file and /run/osbuild/tree/usr/lib64/python3.9/asyncio/__pycache__/tasks.cpython-39.opt-2.pyc, using system_u:object_r:lib_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/77a7a52f5802c21a9b47e04d93fd0b26c364dae425d331bdb71723dcd2babf.file and /run/osbuild/tree/usr/lib64/python3.9/lib2to3/fixes/__pycache__/fix_imports2.cpython-39.opt-2.pyc, using system_u:object_r:lib_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/800f810969582c953c908f6e0a8dbece57ef5897c77bdc0deface215a116c2.file and /run/osbuild/tree/usr/lib64/python3.9/lib2to3/refactor.py, using system_u:object_r:lib_t:s0.
setfiles: conflicting specifications for /run/osbuild/tree/sysroot/ostree/repo/objects/00/ed3d2b2c8f168843eb2d5d99e3f86ddb0d3b62cea9c671a1211ca4548b15ba.file and /run/osbuild/tree/usr/lib64/python3.9/encodings/euc_kr.py, using system_u:object_r:lib_t:s0.
...
```

Seems to be happening on pretty much every file:
```
$ grep -c "conflicting spec" build.log
26151
```

2. Every stage in the deployment pipeline, so every stage we run in the container buidroot, prints the following output:
```
/usr/lib/tmpfiles.d/rpcbind.conf:2: Failed to resolve user 'rpc': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:30: Failed to resolve user 'rpcuser': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:31: Failed to resolve user 'rpcuser': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:32: Failed to resolve user 'rpcuser': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:34: Failed to resolve group 'polkitd'.
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:53: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:55: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:56: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:57: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:58: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:59: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:60: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:61: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:70: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:76: Failed to resolve group 'mail'.
Failed to create directory or subvolume "/usr/local/bin": Read-only file system
Failed to create directory or subvolume "/usr/local/etc": Read-only file system
Failed to create directory or subvolume "/usr/local/games": Read-only file system
Failed to create directory or subvolume "/usr/local/include": Read-only file system
Failed to create directory or subvolume "/usr/local/lib": Read-only file system
Failed to create directory or subvolume "/usr/local/man": Read-only file system
Failed to create directory or subvolume "/usr/local/sbin": Read-only file system
Failed to create directory or subvolume "/usr/local/share": Read-only file system
Failed to create directory or subvolume "/usr/local/src": Read-only file system
ostree admin os-init default --sysroot=/run/osbuild/tree
/usr/lib/tmpfiles.d/rpcbind.conf:2: Failed to resolve user 'rpc': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:30: Failed to resolve user 'rpcuser': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:31: Failed to resolve user 'rpcuser': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:32: Failed to resolve user 'rpcuser': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:34: Failed to resolve group 'polkitd'.
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:53: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:55: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:56: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:57: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:58: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:59: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:60: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:61: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:70: Failed to resolve user 'sssd': No such process
/usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:76: Failed to resolve group 'mail'.
```

**Comment by cgwalters** [comment]
Hmm is the osbuild stage trying to overwrite the labels for the ostree state, i.e. the stuff in `/ostree`?  It really shouldn't, we should just set up a "bootstrap labeling" for `/` and `/ostree`, but thereafter kernel labeling should work for the files outside of the root, and then ostree will write exactly the labels expected for objects and the target root.

**Comment by mvo5** [comment]
Thanks Colin and Achilleas (I'm probably stating the obvious) - the issue seems to be that we run our org.osbuild.selinux stage and we set the file_contexts there:
```
          "type": "org.osbuild.selinux",
          "options": {
            "file_contexts": "etc/selinux/targeted/contexts/files/file_contexts",
```
when I do this: `setfiles -F /etc/selinux/targeted/contexts/files/file_contexts /` inside a `podman run --rm  -it quay.io/centos-bootc/fedora-bootc:eln` environment I get the same errors about conflicting specifications. But AIUI we need to label in our buildroot because the `podman image mount` we do in our `org.osbuild.container-deploy` will label everything to just `unconfined_u:object_r:container_ro_file_t:s0`. I will need to look more tomorrow, it's getting late (and my experience with selinux is still limited sadly).

**Comment by achilleas-k** [comment]
> I will need to look more tomorrow

👍

**Comment by cgwalters** [comment]
> /usr/lib/tmpfiles.d/rpm-ostree-autovar.conf:30: Failed to resolve user 'rpcuser': No such process

We shouldn't be trying to run `systemd-tmpfiles` in general as part of the build process, but ensure things are created on firstboot.  (Though this clashes with things like generating users at build time, but if we can scope that to just running tmpfiles for the necessary `/var/home` prefix that's best)

That said, to fix this type of stuff we currently need to ensure that `altfiles` is set up in the target environment - which it is if we run the OS image as its own installation root!

**Comment by achilleas-k** [comment]
Noticed something while looking into the issues here (and #149) (should've been caught in the pipeline PR in images).
The selinux stage in the build root is running on the container tree (`podman mount`ed and copied out).  The conflicting specifications are (from what I understand) all the hardlinked files that match multiple paths.
The ostree.selinux stage wont help us here (unless we modify it), because that stage mounts the deployment by default and we already have a deployment.

**Comment by cgwalters** [comment]
I spent some time trying to debug and understand this today.  One thing I am struggling with is I don't yet have an efficient edit-compile-debug-cycle with trying to patch Go code in osbuild/images, then rebuild this project and rerun.  And I don't understand osbuild debugging either (the lack of anything like `logrus.Debugf`... I just must be missing a better way to debug?)

I do want to level set again that our existing base images now have (since [this landed](https://github.com/containers/bootc/pull/260)) a single one (ok two) liner to create a correct sparse raw disk image by executing the container itself.  No copying the image around (3? at least?) times, SELinux works correctly, we honor the bootc install config and kargs etc.  (There's no support for injecting external user data or partitions, but that's currently somewhat intentional)

> truncate -s 10G ~/build/centos-bootc.raw
> podman run --pull=newer --pid=host --rm --privileged --security-opt label=type:unconfined_t -ti -v ~/build:/output quay.io/centos-bootc/centos-bootc-dev:stream9 bootc install to-disk --via-loopback /output/centos-bootc.raw

(hmm, I guess we should add the default image size into the container image or its metadata, then we could add `bootc install to-file` which would create the sparse file itself)

This invocation uses the container as the buildroot (like this PR tries to), so it fixes the biggest problem that bib can't target c9s (and a cascade of related problems, like forcing everyone to fetch RPMs just to build the disk image for their container and understanding cache management for those distinct from their container images).

Now of course this raw disk image is just a base thing that still needs significant logic to handle e.g. converting to a format for uploading to AWS, etc. and it doesn't make sense (IMO) to ship that in the OS itself, but in a tool like this.  One thing I briefly tried to dig into a bit more then is how to chain these two into one invocation, which gets into two related problems:

- container-in-container
- accessing containers-storage from a container

Now related to the second, I actually don't see how https://github.com/osbuild/bootc-image-builder/pull/120 can work today without doing something similar to what bootc does for running skopeo from the host mountns, or just trying to bind mount in the container storage (I'd need to ask about best practices for that).

**Comment by cgwalters** [comment]
>  The conflicting specifications are (from what I understand) all the hardlinked files that match multiple paths.

Hmm yes...probably from the ostree stuff (`/sysroot`) in the container image itself.  Probably the simplest work around is to bind mount an emptydir over `/sysroot` for that part?

**Comment by mvo5** [comment]
Fwiw, this should be unblocked by:
https://github.com/osbuild/osbuild/pull/1552 and
https://github.com/osbuild/images/pull/400

Once merged we just need to update the go.mod and the selinux issue should be resolved.

File: Containerfile
Comment (mvo5) [reply Containerfile]: I reverted this now that your PR has landed

File: bib/go.mod
Comment (cgwalters) [inline bib/go.mod:8]: Note this effectively rolls in https://github.com/osbuild/bootc-image-builder/pull/151 - and it seems to be hitting the same linter failure.  I don't understand it.

Comment (mvo5) [reply bib/go.mod:8]: Yeah, I saw this in a different PR of mine as well :/

**Comment by cgwalters** [comment]
Looks like the TMT failure is on selinux denials
> E         Jan 24 16:11:46 3fc17449-b4df-459d-b901-ef10c2b7bfb6.testing-farm audit[21927]: AVC avc:  denied  { nnp_transition nosuid_transition } for  pid=21927 comm="org.osbuild.ost" scontext=system_u:system_r:install_t:s0:c378,c623 tcontext=system_u:system_r:mount_t:s0:c378,c623 tclass=process2 permissive=0

That could actually be new in this PR?

**Comment by cgwalters** [comment]
One thing I guess I should say is that I recently noticed that boot's attempt to detect `install_t` has apparently for a while been not quite working and what's actually been happening is it's just been falling back to `setenforce 0` :angel: 
So that may be one reason what bootc is doing is working and this isn't...

**Comment by cgwalters** [comment]
OK I did https://github.com/containers/bootc/pull/284 and yeah now I'm just in a battle to try to gain install_t again...I would swear this code worked at some point.  I need to trace through whether what's happening here is working and if so how it differs.

**Comment by rhatdan** [comment]
I would like to eliminate  `--security-opt label=type:unconfined_t `

**Comment by mvo5** [comment]
> Looks like the TMT failure is on selinux denials
> 
> > E         Jan 24 16:11:46 3fc17449-b4df-459d-b901-ef10c2b7bfb6.testing-farm audit[21927]: AVC avc:  denied  { nnp_transition nosuid_transition } for  pid=21927 comm="org.osbuild.ost" scontext=system_u:system_r:install_t:s0:c378,c623 tcontext=system_u:system_r:mount_t:s0:c378,c623 tclass=process2 permissive=0
> 
> That could actually be new in this PR?

Thank you, yes, that is a regression. I think i know what is wrong and opened https://github.com/osbuild/images/pull/402

**Comment by achilleas-k** [comment]
so [not a caching issue](https://github.com/osbuild/bootc-image-builder/actions/runs/7658097303/job/20870298251?pr=138) :roll_eyes:

**Comment by achilleas-k** [comment]
golangci-lint is failing for unknown reasons.  Let's force merge this and see if it also happens on the merge queue after rebase.

**Review by achilleas-k — APPROVED** [review]

**Comment by mvo5** [comment]
Fwiw, the changes from @achilleas-k look good to me (thank you for your help here!)

**Comment by achilleas-k** [comment]
> golangci-lint is failing for unknown reasons. Let's force merge this after everything else succeeds and see if it also happens on the merge queue after rebase.

I dropped the linter requirement from the merge rules instead.  I'll enable it again later.

**Pull request merged**
